File: Codes/Level 2/Trees/BinaryTreeZigZagOrderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        if(root == NULL)    return ans;
        deque<TreeNode*> q;
        int counter = 0;
        q.push_back(root);
        while(q.size()){
            int k = q.size();
            vector<int> currAns;
            while(k--){
                TreeNode* curr = q.front();
                q.pop_front();
                if(curr->left)  q.push_back(curr->left);
                if(curr->right) q.push_back(curr->right);
                currAns.push_back(curr->val);
            }
            if(counter++ % 2)   reverse(currAns.begin(), currAns.end());
            ans.push_back(currAns);
        }
        return ans;
    }
};

File: Codes/Level 2/Trees/BinaryTreeZigZagOrderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if(root == null)    return ans;
        Deque<TreeNode> q = new ArrayDeque<>();
        int counter = 0;
        q.addLast(root);
        while(q.size() > 0){
            List<Integer> currAns = new ArrayList<>();
            int k = q.size();
            while(k-- > 0){
                TreeNode curr = q.getFirst();
                q.removeFirst();
                if(curr.left != null)   q.addLast(curr.left);
                if(curr.right != null)  q.addLast(curr.right);
                currAns.add(curr.val);
            }
            ans.add(currAns);
            if(counter++%2 == 1) Collections.reverse(currAns);
        }
        return ans;
    }
}